src/v/wasm/allocator.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/aligned_buffer.hh>
+#include <seastar/core/chunked_fifo.hh>
+#include <seastar/util/optimized_optional.hh>
+
+#include <type_traits>
+
+namespace wasm {
+
+/**
+ * An owned bit of heap memory, which is aligned to page size and it's size is a
+ * multiple of the page size.
+ */
+struct heap_memory {
+    std::invoke_result_t<
+      decltype(ss::allocate_aligned_buffer<uint8_t>),
+      size_t,
+      size_t>
+      data;
+    size_t size;
+};
+
+// The allocator for "heap" memory within a Wasm VM.
+//
+// This "heap" memory is known as LinearMemory in WebAssembly parlance and must
+// be a contiguous range of memory. Since these memories can grow beyond the
+// maximum amount of memory we recommend allocating at once in Redpanda (due to
+// fragmentation in long running processes), we allocate a pool at startup when
+// there are copious amounts of larger contiguous chunks.
+//
+// Additionally memory must be page-aligned and a multiple of page size to
+// enable performance optimizations within the Wasm VMs.
```

Comment:
does wasmtime also do stuff like mprotect to mark pages executable?

i guess if we are changing protection on memory then my concern would be oh if we change protection on some allocated memory form the seastar allocator could we unknowingly change protection on some memory unassociated with the page because of the way the seastar allocator divides up memory?

it seems safe with the alignment and size constraints, but might be worth checking with travis/perf-team if you haven't already.

Reply:
> does wasmtime also do stuff like mprotect to mark pages executable?

Yes and no. Not in the code paths we're going to use, and not this memory.

This memory is the data memory usage - executable memory is yet another pool that will be left outside of seastar for now.

> it seems safe with the alignment and size constraints, but might be worth checking with travis/perf-team if you haven't already.

mprotect is not called on the memory here (one of my next PRs will do some stuff around this tho) is there anything I should call to their attention?

Reply:
>mprotect is not called on the memory here (one of my next PRs will do some stuff around this tho) is there anything I should call to their attention?

i don't think so.